### PR TITLE
[PA][Gemma4] Fix prefix caching for bidirectional attention path

### PR DESCRIPTION
**Pull request by p-wysocki**
### Details:
 - Gemma models let the tokens that make up an image attend to each other in both directions
 - For prefix caching case the start and end positions of each image region were recorded in the wrong coordinate system, so once part of a prompt had already been cached the region was located in the wrong spot
 - This change lets image tokens attend across their entire image and fixes the region positions so results stay correct with prefix caching and batched requests
 - Pure text generation is unchanged

### Tests:
 - `prompt_lookup=True` tests in GenAI pass for me locally. Also the tests skipped in GenAI's `OPTIMUM_VS_GENAI_MODEL_EXPECTED_FAIL_CASES` pass for me - both tested with `tiny-random-gemma4-moe`
 - I ran a Gemma4 with text+image inputs to verify the output correctness
 - Ran one-shot prefill comparison of Gemma4 vs HF PyTorch MoE, so the one-shot results are correct
 - Added chunked prefill vs one-shot prefill comparison tests for PA op, verifying that chunked prefill matches the correct one-shot results

### Tickets:
 - CVS-189726

### AI Assistance:
 - *AI assistance used: yes*
 - Debugging the issue, adding tests, the AI's contribution has been verified

### CODE REVIEW

File: src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/x64/paged_attn_token_type.cpp [hunk 1]
```diff
@@ -242,6 +242,121 @@ class PagedAttnTokenTypeTest : public testing::WithParamInterface<PagedAttnToken
...
+    RunResult run_pa_chunked(std::shared_ptr<ov::Model> model,
+                             ov::element::Type data_type,
+                             size_t seq_len,
+                             size_t head_size,
+                             size_t head_num,
+                             const std::vector<int32_t>& token_types,
+                             size_t chunk1_len) {
+        OPENVINO_ASSERT(token_types.size() == seq_len);
+        OPENVINO_ASSERT(chunk1_len > 0 && chunk1_len < seq_len);
+
```
[the remark this answers is not in the document]
Comment (p-wysocki) [reply src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/x64/paged_attn_token_type.cpp:258]: Added an explicit assert to restrict the helper to f32.

File: src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/x64/paged_attn_token_type.cpp [hunk 2]
```diff
@@ -453,6 +568,41 @@ TEST_P(PagedAttnTokenTypeTest, ImageTokensWithSlidingWindowDifferFromCausal) {
...
+    const size_t seq_len = pattern.types.size();
+
+    size_t first_image_pos = seq_len;
+    for (size_t i = 0; i < seq_len; ++i) {
+        if (pattern.types[i] == 1) { first_image_pos = i; break; }
+    }
+    ASSERT_GT(first_image_pos, 0u) << "Pattern must start with a text token";
+    ASSERT_LT(first_image_pos, seq_len) << "Pattern must contain an image token";
+    const size_t chunk1_len = first_image_pos;
```
[the remark this answers is not in the document]
Comment (p-wysocki) [reply src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/x64/paged_attn_token_type.cpp:766]: Added `ChunkedPrefillBatchedMatchesSingleShot` test.

**Review by mitruska — APPROVED** [review]

**Review by yatarkan — APPROVED** [review]

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp [hunk 3]
```diff
@@ -465,8 +465,8 @@ struct MHAHelper {
 
     // Precompute image group boundaries from token_type_ids.
     // For each image token:
-    //   _image_group_end[i] = index past the last contiguous image token in the same group
```
Comment (zhangYiIntel) [inline src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp:468]: I think we could elaborate this comment here. Even though the formula is correct but we could make it clear that:
```
For sequence seq:

current chunk token index is in global batched-token space: i
chunk-local index is: i - seq_begin
KV-global index should be: past_len[seq] + (i - seq_begin)

That is why the code uses:

kv_offset = past_len[seq] - seq_begin
group_begin_kv = group_begin + kv_offset
```

File: src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/x64/paged_attn_token_type.cpp [hunk 4]
```diff
@@ -242,6 +242,300 @@ class PagedAttnTokenTypeTest : public testing::WithParamInterface<PagedAttnToken
...
+
+        auto run_chunk = [&](size_t row0, size_t nrows, int32_t past_len, int32_t nblocks) {
+            ov::Tensor q = slice_rows(q_full, row0, nrows);
+            ov::Tensor k = slice_rows(k_full, row0, nrows);
+            ov::Tensor v = slice_rows(v_full, row0, nrows);
+
+            ov::Tensor past_lens(ov::element::i32, {1});
+            ov::Tensor subsequence_begins(ov::element::i32, {2});
+            ov::Tensor block_indices_begins(ov::element::i32, {2});
+            past_lens.data<int32_t>()[0] = past_len;
```
Comment (zhangYiIntel) [inline src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/x64/paged_attn_token_type.cpp:322]: maybe we should cover another corner case, if a block is filled part of text tokens and part of img tokens for one sequence

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp (hunk 3 above)
Comment (p-wysocki) [reply src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp:468]: Good idea. I modified the comment to include the formula, I hope it's clearer now.

File: src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/x64/paged_attn_token_type.cpp (hunk 4 above)
Comment (p-wysocki) [reply src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/x64/paged_attn_token_type.cpp:322]: I added new test `mix_across_blocks`, where:

- [Block 0] tokens 0-15 are text tokens
- [Block 0] tokens 16-31 are image tokens
- [Block 1] tokens 32-35 are image tokens, spilled from block 0
- [Block 1] tokens 35-47 are text tokens

**Review by zhangYiIntel — APPROVED** [review]
